### PR TITLE
parse unencrypted results for confidential txs

### PR DESCRIPTION
**Pull request by Andrew7234**
https://app.clickup.com/t/8692tq704

```
oasisindexer=> select count(*) from chain.runtime_transactions where evm_encrypted_format='encrypted/x25519-deoxysii' and method='evm.Call' and success=true and evm_encrypted_result_data is not NULL;
 count
-------
 15318 <-- encrypted results
(1 row)

oasisindexer=> select count(*) from chain.runtime_transactions where evm_encrypted_format='encrypted/x25519-deoxysii' and method='evm.Call' and success=true and evm_encrypted_result_data is NULL;
 count
--------
 212680 <-- unencrypted results
```

Before:
```
indexer=# select runtime, round, tx_index, method, success, evm_encrypted_result_data from chain.runtime_transactions
;
 runtime  |  round  | tx_index |  method  | success | evm_encrypted_result_data
----------+---------+----------+----------+---------+---------------------------
 sapphire | 2993605 |        0 | evm.Call | t       |
 sapphire | 2993605 |        1 | evm.Call | t       |
 sapphire | 2993606 |        0 | evm.Call | t       |
 sapphire | 2993606 |        1 | evm.Call | t       |
 sapphire | 2993607 |        0 | evm.Call | t       |
 sapphire | 2993607 |        1 | evm.Call | t       |
 sapphire | 2993608 |        0 | evm.Call | f       |
(7 rows)
```

After
```
indexer=# select runtime, round, tx_index, method, success, evm_encrypted_result_data from chain.runtime_transactions
;
 runtime  |  round  | tx_index |  method  | success |          evm_encrypted_result_data
----------+---------+----------+----------+---------+----------------------------------------------
 sapphire | 2993605 |        0 | evm.Call | t       | \x76115a6703f475bdc9eee0747ba441cedc4ed9ebfe
 sapphire | 2993605 |        1 | evm.Call | t       | \x7c495d2602e702b7b490779a5e029f436fcfb02c7b
 sapphire | 2993606 |        0 | evm.Call | t       |
 sapphire | 2993606 |        1 | evm.Call | t       | \x40d51310f0ab71197168e70d45b3bff595c10840d5
 sapphire | 2993607 |        0 | evm.Call | t       | \xa8eafc2f40b431323d3017142d8c87208f8e5d8d28
 sapphire | 2993607 |        1 | evm.Call | t       | \x74c6ed9bd74cdbfeec89c4c55e3ada66dc94f2b8b5
 sapphire | 2993608 |        0 | evm.Call | f       |
(7 rows)
```

On testnet, after block 542000 we stopped parsing the tx results correctly... will need a full reindex.

Future Todo: Make sure this is covered in the e2e_regression tests when we add sapphire to them.

### CODE REVIEW

File: analyzer/runtime/evm/client.go
Comment (mitjat) [inline analyzer/runtime/evm/client.go:339]: Heads up: I've noticed this error in logs quite a bit:
```
{
analyzer: "sapphire"
caller: "extract.go:499"
err: "outer call format plain (0) not supported"
level: "error"
module: "analysis_service"
msg: "error unmarshalling encrypted data and result, omitting encrypted fields"
}
```

The `caller` above is a line in `extract.go` that calls the method from this PR (`EVMMaybeUnmarshalEncryptedData`). So basically, we've got txs with `call.Format == sdkTypes.CallFormatPlain`, and we're not handling them. (This function generates the same error message twice, once here and once in line 309.)
I haven't created a ticket for this IIRC, but I'm wondering if it's on your radar already since you've been looking at encryption a lot lately? It's probably worth handling in the next days as you're working in this area.

File: analyzer/runtime/evm/client.go
Comment (mitjat) [inline analyzer/runtime/evm/client.go:326]: Which of `callResult.IsSuccess()` and `!callResult.IsUnknown()` indicates unencrypted input? (I'm referring to your comment.)

Can a `callResult` have `IsSuccess()` and `IsUnknown()` at the same time? If not, let's just check for `IsSuccess()` here. Edit: Urgh I see how those two methods are defined. Not nice. Worth commenting on here, I think.

Why do we handle the case of tx status=failed (as opposed to status={ok, unknown}) in such a different place, [here](https://github.com/oasisprotocol/nexus/blob/5881904c1f9a6de691dabda76152f33c0d98d7f5/analyzer/runtime/extract.go#L344-L344)? Are error messages never encrypted? Also worth commenting on here..

**Review by mitjat — COMMENTED** [review]
Thank yo ufor making sense of this.
I only have comments/questions that will hopefully make it make sense to myself and future maintainers.

**Comment by mitjat** [comment]
> On testnet, after block 542000 we stopped parsing the tx results correctly... will need a full reindex.

Huh. How come we were parsing them correctly before? / What caused us to start indexing incorrectly?
Also, this is about the production deployment, presumably?

File: analyzer/runtime/evm/client.go
Comment (Andrew7234) [reply analyzer/runtime/evm/client.go:339]: Thanks for raising this! I think in this case, the analyzer is working as expected since we aren't able to extract `evm_encrypted_*` fields from a plain tx. I'll add an explicit switch case to reduce the log spam here though.

The one equivalent field we _could_ extract is the `data_data`, eg the evm call data. However, it likely requires its own dedicated column, and also we currently store that as part of the `body -> data`.

File: analyzer/runtime/evm/client.go
Comment (pro-wh) [inline analyzer/runtime/evm/client.go:323]: why is this changed to 'ok'? we're still unmarshalling the .Unknown

File: analyzer/runtime/evm/client.go
Comment (pro-wh) [reply analyzer/runtime/evm/client.go:326]: `!callResult.IsUnknown()` should be covered by the above case not matching

File: analyzer/runtime/evm/client.go
Comment (pro-wh) [reply analyzer/runtime/evm/client.go:339]: wait what do you mean by "twice" and "log spam"? it shows up in the code lexically twice, but they return when either one happens

**Comment by Andrew7234** [comment]
> Huh. How come we were parsing them correctly before? / What caused us to start indexing incorrectly?

Hmm iirc there was some talk of how the data in `callResult.Unknown` was moved to `callResult.Ok` for encrypted txs, but I'll need to follow up on that

> Also, this is about the production deployment, presumably?

Ah it's actually staging testnet

File: analyzer/runtime/evm/client.go
Comment (pro-wh) [reply analyzer/runtime/evm/client.go:326]: I need to read through this thread more tho

File: analyzer/runtime/visitors.go
Comment (pro-wh) [reply analyzer/runtime/visitors.go:114]: oops thanks

File: analyzer/runtime/evm/client.go
Comment (Andrew7234) [reply analyzer/runtime/evm/client.go:323]: fixed!

File: analyzer/runtime/evm/client.go
Comment (pro-wh) [reply analyzer/runtime/evm/client.go:329]: ok I don't get the code comment here either.

I thought the .Unknown, .Ok, and .Failed are mutually exclusive. Unknown is allowed by the SDK, but I think I heard that Sapphire moved to making the successfulness of transactions public, meaning it chooses to use .Ok and .Failed instead of .Unknown.

Additionally, I think the implication in this PR is that .Failed is always in cleartext in Sapphire? Is that the case? Is there a systematic way to tell whether .Failed is encrypted or not?

Comment (pro-wh) [reply analyzer/runtime/evm/client.go:329]: https://github.com/oasisprotocol/oasis-sdk/blob/runtime-sdk/v0.7.4/runtime-sdk/src/callformat.rs#L253 yes it look like failure gets passed through unencrypted under this `expose_failure` setting

Comment (pro-wh) [reply analyzer/runtime/evm/client.go:329]: ok correction: it's not Sapphire specifically doing this, newer versions of the SDK do this for the whole EVM module https://github.com/oasisprotocol/oasis-sdk/blob/runtime-sdk/v0.7.4/runtime-sdk/modules/evm/src/lib.rs#L651

File: analyzer/runtime/evm/client.go
Comment (Andrew7234) [reply analyzer/runtime/evm/client.go:342]: Hm so when the outer call format is `plain (0)`, do we still want to log an error? My original idea was to simply return no error in this case, since the docstring says that `If the data is not encrypted,it returns nil with no error.`. I might be missing something though. If the callformat is 0, does it guarantee that the transaction is not encrypted?

Comment (pro-wh) [reply analyzer/runtime/evm/client.go:342]: "if the data is not encrypted" here refers to a certain thing about how Sapphire style encryption is. the data field of an otherwise unencrypted Ethereum format transaction is a valid CBOR-encoded Oasis `Call` structure (the hope is that no contract will incidentally expect anything compatible as the actual data, initcode for "create" as well).

so the phrase in the description (as well as the "Maybe" in the name) there refers to handling this ambiguity. it's only to say you can pass in whatever the data/initcode you get and it'll do the right thing.

as for how plain (0) should fit in, imo return nil with no error. the goal of this function is to populate the encrypted envelope fields for insertion into the database. and the change in this PR to look for the encrypted envelope in callResult.Ok is correct too.

Comment (pro-wh) [reply analyzer/runtime/evm/client.go:342]: addendum: now that I think about it, maybe it's more suitable to handle invalid `result` CBOR with similar logic inside this function instead of special-casing the code around it to pass `nil` for "create."

File: analyzer/runtime/evm/client.go
Comment (aefhm) [reply analyzer/runtime/evm/client.go:334]: Agree with @pro-wh

Comment (aefhm) [reply analyzer/runtime/evm/client.go:334]: Oof, I suppose the history is tough because we need to support pre-SDK change where errors and successes were both in `Unknown`.

Comment (pro-wh) [reply analyzer/runtime/evm/client.go:334]: pre-change will be fine, we continue to look in Unknown if it's Unknown

Comment (mitjat) [reply analyzer/runtime/evm/client.go:334]: > My current understanding is that if the inner sdk callResult fails, that error is bubbled up to the outer evm callResult.Failed and so we can handle failed txs in one place at the higher level

This is a good comment, can you add it as a code comment to the `default:` branch on line 338 (which is where we handle errors)?

By `callResult.Failed`, do you mean `txr.Failed` as checked [here](https://github.com/oasisprotocol/nexus/blob/5881904c1f9a6de691dabda76152f33c0d98d7f5/analyzer/runtime/extract.go#L344-L344)?

Regardless, I'm still a little confused by the flow: If the transaction was encrypted and failed, we'll detect the failure in the `default:` branch of `case` (line 338), we'll return an error, but our caller is here and it just logs a parsing failure.
https://github.com/oasisprotocol/nexus/blob/9e33faed3fc86d90edd16892b51826a38cc569fc/analyzer/runtime/extract.go#L514-L517
Should we set `chain.runtime_transactions.success`? At the very least, we probably shouldn't be logging a parsing error.

**Review by mitjat — APPROVED** [review]
Thank you. Approving so it can be merged when convenient. The first comment thread still needs more work though, or I'm misunderstanding something.

File: analyzer/runtime/evm/client.go
Comment (Andrew7234) [reply analyzer/runtime/evm/client.go:347]: yup!

File: analyzer/runtime/evm/client.go
Comment (Andrew7234) [reply analyzer/runtime/evm/client.go:334]: confirmed that when the tx fails, the parsing error isn't reached since `VisitCall` sets result to nil. I'll add a coment

Comment (pro-wh) [reply analyzer/runtime/evm/client.go:334]: it could be unknown outer + fail inner for non-evm transactions and for old transactions

**Comment by mitjat** [comment]
IIRC you mentioned in Split that the logic likely needs to fixed in some way still?

I don't remember the conclusion in terms of actual code, but I do remember we said it would be useful to document the expected layout (i.e. contents of outer and inner result) for all the combos:
{old format, new format} x {success, failure} x {encrypted, "encrypted" with Plain format, unencrypted}

It's a lot of them, but that's also the reason we're having such long conversations, and lower-than-ideal confidence in correctness. If we can make meaningful unit tests for them, even better. (That's kinda scope creep for the PR, so feel free to push back. But it's only _kinda_ scope creep, and it covers very real existing tech/documentation debt.)

If you don't know the answer for some of the combos, we can produce that combo on testnet, or find it in the wild, or ask around.

As I've felt several times with this PR: Thank you for your service 🫡

File: analyzer/runtime/evm/client.go
Comment (Andrew7234) [inline analyzer/runtime/evm/client.go:343]: per pro-wh's comment:
> it could be unknown outer + fail inner for non-evm transactions and for old transactions

Adjusted to extract the `FailedCallResult` and bubble it up to the calling fn in `extract.go`, which in turn uses the failed call result to populate the error fields.

**Comment by Andrew7234** [comment]
Ok added tests and confirmed that they work as expected. I still need to locate some older sapphire encrypted txs that have the tx/results wrapped in the sapphire outer call within the oasis outer call, which requires archive node access

**Review by mitjat — APPROVED** [review]
We were discussing this ticket with Tadej during sprint planning, and the one thing he immediately said when he learned about your [writeup in Google Docs](https://docs.google.com/document/d/1LHb94T0i9IpT_zI6WU80kiErtT7T1asvdOsKrcPoNE0/edit) was how nice it would be to have it close to the codebase so it doesn't get lost.

You did add nice inline code comments but they serve more so the function of "now that we know how the data is laid out, the comments guide us through how the nested ifs correspond to that layout".
I too would like to encourage you to include the contents of the Google Doc (with links converted to Markdown(?), and the table rendered as nested bullet points) in a place where a) it's discoverable when browsing the code and b) it's accessible to the public. It would be a shame if the results of days of digging and summarizing end up rotting away in some forgotten doc.

Some placement options:
- Huge docstring for `EVMMaybeUnmarshalEncryptedData()`
- A separate markdown file (in this directory?), referenced prominently from the docstring of `EVMMaybeUnmarshalEncryptedData()`.

I'm re-LGTMing -- this feels well researched now, and even if it still fails in some corner case, we can deal with that later. I bet it will feel good to move on :). Thank you!

File: analyzer/runtime/extract_test.go
Comment (Andrew7234) [reply analyzer/runtime/extract_test.go:40]: Good eye - I still need to find some old txs that store the failed result in the `outer sapphire call result`. Will add shortly

File: analyzer/runtime/evm/client.go
Comment (Andrew7234) [reply analyzer/runtime/evm/client.go:77]: Added! Updated to 
```
        // Unlike `ResultData` and `DataData`, this field is not encrypted data. Rather,
	// it is extracted during encrypted tx parsing and processed in extract.go.
	// If non-null, the tx is an older Sapphire tx and it failed, and `ResultNonce`
	// and `ResultData` will be empty.
```

Comment (mitjat) [reply analyzer/runtime/evm/client.go:77]: > If non-null, the tx is an older Sapphire tx

Oh. So where do we store failure details for all other transactions? (notably, new Sapphire)

Comment (Andrew7234) [reply analyzer/runtime/evm/client.go:81]: It's stored in the "outer call result" one level up

File: analyzer/runtime/extract_test.go
Comment (Andrew7234) [reply analyzer/runtime/extract_test.go:40]: Ok I couldn't find any - will keep searching and add in a follow on PR

**Pull request merged**
